libs/content/workspace/src/workspace.rs

```diff
@@ -540,24 +540,16 @@ impl Workspace {
                     let core = self.core.clone();
                     let update_tx = self.updates_tx.clone();
                     let ctx = self.ctx.clone();
-                    let seq = if let Some(TabContent::Markdown(md)) = &tab.content {
-                        md.editor.buffer.current.seq
-                    } else {
-                        0
-                    };
 
                     thread::spawn(move || {
-                        let content = save_req.content;
-                        let id = save_req.id;
-                        let hmac = save_req.hmac;
+                        let SaveRequest { seq, content, id, old_hmac, safe_write } = save_req;
 
-                        let result = if hmac.is_some() {
-                            core.safe_write(id, hmac, content.clone().into())
-                                .map(|hmac| (content, Some(hmac), Instant::now(), seq))
+                        let result = if safe_write {
+                            core.safe_write(id, old_hmac, content.clone().into())
+                                .map(|new_hmac| (content, Some(new_hmac), Instant::now(), seq))
                         } else {
-                            // todo: what if editor opened a file with no hmac, then we overwrite outside edits here
                             core.write_document(id, content.as_bytes())
-                                .map(|_| (content, hmac, Instant::now(), seq))
+                                .map(|_| (content, None, Instant::now(), seq))
```

Comment:
In the old version, where the variables here were all named `hmac`, an unsafe write would pass on the old hmac in place of the new hmac which would have been returned by `core.safe_write()`. More descriptive variable names highlighted this issue.

This isn't related to the bug, though, because it causes the `SaveResult` to contain a `None` hmac on the first write which is the correct behavior.